app/app/vhosts/edit/route.js

```diff
@@ -1,55 +1,99 @@
+    let stack = app.get('stack');
+
+    return Ember.RSVP.hash({
+      vhost: this.store.find('vhost', params.vhost_id),
+      certificates: stack.get('certificates'),
+      stack: stack
+    });
+  },
+
+  afterModel(model) {
+    let vhost = model.vhost;
+
+    return Ember.RSVP.hash({
+      service: vhost.get('service'),
+      certificate: vhost.get('certificate')
+    });
+  },
+
+  setupController(controller, model) {
+    let vhost = model.vhost;
+    let certificates = model.certificates;
+
+    controller.set('model', vhost);
+    controller.set('certificates', model.certificates);
+    controller.set('vhostCertificate', certificates.objectAt(0));
+    controller.set('showAddNewCertificate', model.certificates.get('length') === 0);
   },
 
   actions: {
-    save: function(){
+    save() {
       let controller = this.controller;
       let app = this.modelFor('app');
-      let vhost = this.currentModel;
+      let vhost = this.currentModel.vhost;
+      let certificatePromise;
 
-      let certificate = controller.get('newCertificate');
-      let privateKey = controller.get('newPrivateKey');
+      // If certificate create UI is visible, create a certificate
+      // If not, use the selected certificate
+      if(controller.get('showAddNewCertificate')) {
+        let stack = this.currentModel.stack;
+        let certificateBody = controller.get('certificateBody');
+        let privateKey = controller.get('privateKey');
+        let certParams = { certificateBody, stack, privateKey };
+        let newCertificate = this.store.createRecord('certificate', certParams);
 
-      vhost.setProperties({
-        certificate,
-        privateKey,
-        app
-      });
-      vhost.save().then( () => {
+        certificatePromise = newCertificate.save();
+      } else {
+        certificatePromise = new Ember.RSVP.resolve(vhost.get('certificate'));
+      }
+
+      certificatePromise.then((certificate) => {
+        vhost.setProperties({ app, certificate });
+        vhost.set('certificateBody', '');
+        vhost.set('privateKey', '');
+        return vhost.save();
+      }).then(() => {
         let op = this.store.createRecord('operation', {
-          type: 'reprovision',
-          certificate,
-          privateKey,
+          type: 'provision',
           vhost
         });
         return op.save();
       }).then( () => {
+        let message = `Domain ${vhost.get('virtualDomain')} updated.`;
         this.transitionTo('app.vhosts.index');
+        Ember.get(this, 'flashMessages').success(message);
+
       }).catch( (e) => {
-        if (e instanceof DS.InvalidError) {
-          // no-op
-        } else {
-          throw e;
-        }
+        let message = Ember.get(e, 'responseJSON.message') || `There was an error updating ${vhost.get('virtualDomain')}`;
+        Ember.get(this, 'flashMessages').danger(message);
       });
     },
 
-    cancel: function(){
+    showAddNewCertificate(show) {
```

Comment:
Seems like this should be on the controller (unless I am mistaken this is for controlling UI only).